### ISSUE TITLE
GraphQLResolver<T> does not resolve fetcher function and throws TypeResolverError

### ISSUE THREAD

**Issue by mhornung85** [graphql-java-kickstart/graphql-java-tools#232]
Hi everybody and thanks for the effort you put into this library.

I´m facing two problems that showed up after I did a slight change to my data class two improve perfomance: I changed two eagerly loaded attributes to lazy - here´s the excerpt of the data class:

`public abstract class Task extends {
	@Id
	public Long id;

	@ManyToOne(fetch = FetchType.LAZY)
	protected Order<? extends Unit> order;

	@OneToOne(fetch = FetchType.LAZY, cascade = { CascadeType.ALL })
	private Address address;
`

two subclasses are implementing this abstract Task, the subclasses are added to the Dictionary.

1. Problem: When I´m trying to request data where the Order is inlined, like: 

`    taskFilter: {date: "11.02.2019"}) {
    id,
    order{
      id
    }
`

I get the error: 
com.coxautodev.graphql.tools.TypeResolverError: Expected object type with name 'Order_$$_jvstaa3_6' to implement interface 'Order', but it doesn't!
at com.coxautodev.graphql.tools.DictionaryTypeResolver.getType(DictionaryTypeResolver.kt:31) ~[graphql-java-tools-5.4.1.jar:na]

I thought its maybe a problem with the proxy class, but the orher relation (Address) works like a charme - even if its a proxy class. Unfortuantely I´m not able to strep through the DictionaryTypeResolver.kt class - I tried for others to inspect the variables inside the kotlin code, but I didn´t get it to work.

2. Problem:
I thought: doesn´t matter, I want to implement a Dataloader for this attribute anyway - so lets just write a custom datafetcher that - if it´s really a problem with the Proxy class - unwraps the class or does anything else and gives back the correct data: 

`@Component
public class TaskDataResolver implements GraphQLResolver<Task>{
	
	@Autowired
	private TaskService taskService;
	
	public Order getOrder(Task task) {
	        ChildOrder childOrder = new ChildOrder(); // ChildOrder is a Sub class of Order (just an example)
		childOrder.setId(1L);
		return childOrder;
	}
	
	public Address getAddress(Task task) {
		Address address = new Address();
		address.setCity("Frankfurt");
		return address;
	}

}`

I tried it with the child class "HandoverTask.class" as well - but graphql-java-tools does not find those resolver functions and does not call it. It falls back - in any case - to the data class getter method or the direct attribute, if I delete the getter method.

To sum up: I don´t have a clue whats the problem here: I expect its the exactly same for the two problems. Something related to inheritance?

Beste regards and thanks
Marius

**Comment by oliemansm** [comment]
Hmm, could be something with inheritance and maybe the combination with generics that I see at the top. Both should work nowadays, but perhaps combining that with proxies is causing the issue now. I see in the resolver you're returning an `Order` without generics, while at the top in the `@ManyToOne` field definition it has generics. When creating the dictionary GraphQL Tools will store `Order<Foo>` and `Order<Bar>` as separate entries. So if you're now returning just `Order` I'm not sure what GraphQL Tools will map that too. Unless in all resolvers GraphQL Tools only knows one particular `Order` without any generics.

So despite all the info and examples you've already given there seems to be missing something more to be able to help out. Is it possible to share your repository showing this problem?

**Comment by mhornung85** [comment]
Hi oliemansm,

thanks for the answer. I tried to create a rep already, but I had to overtake the codebase a few weeks a go and its full of glue code and i cannot get it work in a simple example in an fair amount of time.

I tried several versions with the Order and the preceding generic: <Unit> <? extends Unit> or without generic. 

Anyway - I took it as an opportunity to get rid of all those weird generics (someone was too lazy to refactor the code properly and then used generics to glue things together) and cleaned up the whole code base. 

I close this for now, but I put it on my bucket list to provide a simple example from scratch to figure out if something is wrong (or maybe not implemented yet) here. . I reopen it then.

**Comment by oliemansm** [comment]
Ok, thanks!

**Comment by kaikun213** [comment]
@mhornung85 any news on this? I ran into the same issue.

In my case I use Spring envers and the returned revision entity has 
`@ManyToOne(FetchType=EAGER)` mappings. E.g.
```
@Audited
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@Data @NoArgsConstructor @EqualsAndHashCode(callSuper = true)
public class Foo {

    @ManyToOne(fetch= FetchType.EAGER)
    @Nonnull
    protected Bar bar;

...
}
```


However, it seems to still be a Hibernate Proxy and then the type resolver throws the same error.


```
Caused by: com.coxautodev.graphql.tools.TypeResolverError: Expected object type with name 'Bar' to implement interface 'Bar', but it doesn't!
	at com.coxautodev.graphql.tools.DictionaryTypeResolver.getType(DictionaryTypeResolver.kt:31) ~[graphql-java-tools-5.7.1.jar:na]
	at graphql.execution.ResolveType.resolveTypeForInterface(ResolveType.java:49) ~[graphql-java-13.0.jar:na]
	at graphql.execution.ResolveType.resolveType(ResolveType.java:27) ~[graphql-java-13.0.jar:na]
	at graphql.execution.ExecutionStrategy.resolveType(ExecutionStrategy.java:684) ~[graphql-java-13.0.jar:na]
	at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:437) ~[graphql-java-13.0.jar:na]
	at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:388) ~[graphql-java-13.0.jar:na]
...
```

**Comment by kaikun213** [comment]
Btw the workaround that I use is to expicitly initialize the fields. It is not pretty, but it works.

E.g. 

```
    public static <T> T initializeAndUnproxy(T entity) {
        if (entity == null) {
            return null;
        }

        Hibernate.initialize(entity);
        if (entity instanceof HibernateProxy) {
            entity = (T) ((HibernateProxy) entity).getHibernateLazyInitializer()
                    .getImplementation();
        }
        return entity;
    }
```
And within an entity I call it for the required fields:
```
    public void initLazyFields() {
        super.initLazyFields();
        setBar(initializeAndUnproxy(getBar()));
    }
```
